File: ios/Packages/PrayerKit/Sources/PrayerKit/PrayerEngine.swift
import Adhan
import Foundation

/// User-facing prayer names. `sunrise` is a displayed time but not a prayer.
public enum PrayerName: String, CaseIterable, Codable, Sendable {
    case fajr, sunrise, dhuhr, asr, maghrib, isha

    public var isPrayer: Bool { self != .sunrise }
}

/// Resolved calculation settings (docs/features/prayer.md). Defaults flow:
/// user override → /v1/config/prayer-defaults(country) → bundled fallback.
public struct PrayerSettings: Codable, Equatable, Sendable {
    public var method: String
    public var madhab: String
    /// nil = automatic: no rule below the high-latitude threshold, recommended
    /// rule above it (spike finding: never rely on library defaults ≥48°).
    public var highLatitudeRule: String?
    /// Manual per-prayer adjustments in minutes (±30 enforced here).
    public var adjustments: [PrayerName: Int]
    public var hijriOffsetDays: Int

    public static let highLatitudeThreshold = 48.0

    public init(
        method: String = "mwl",
        madhab: String = "shafi",
        highLatitudeRule: String? = nil,
        adjustments: [PrayerName: Int] = [:],
        hijriOffsetDays: Int = 0
    ) {
        self.method = method
        self.madhab = madhab
        self.highLatitudeRule = highLatitudeRule
        self.adjustments = adjustments.mapValues { max(-30, min(30, $0)) }
        self.hijriOffsetDays = max(-2, min(2, hijriOffsetDays))
    }
}

/// One civil day's times, adjustments applied.
public struct PrayerDay: Equatable, Sendable {
    public let date: DateComponents // y/m/d civil date at the location
    public let times: [PrayerName: Date]

    public func time(_ name: PrayerName) -> Date { times[name]! }

    /// Ordered for display.
    public var ordered: [(name: PrayerName, time: Date)] {
        PrayerName.allCases.map { ($0, times[$0]!) }
    }
}

public struct NextPrayerState: Equatable, Sendable {
    /// The prayer whose window we are in (nil before Fajr).
    public let current: PrayerName?
    public let next: PrayerName
    public let nextTime: Date
}

/// Deterministic prayer engine over PrayerCalculator. Pure: all inputs explicit,
/// no clock or location reads — callers own those effects.
public struct PrayerEngine: Sendable {
    private let calculator = PrayerCalculator()

    public init() {}

    public func day(
        latitude: Double,
        longitude: Double,
        date: DateComponents,
        settings: PrayerSettings
    ) throws -> PrayerDay {
        let rule = Self.effectiveHighLatitudeRule(settings: settings, latitude: latitude)
        let raw = try calculator.times(for: .init(
            latitude: latitude,
            longitude: longitude,
            date: date,
            method: settings.method,
            madhab: settings.madhab,
            highLatitudeRule: rule
        ))
        var times: [PrayerName: Date] = [
            .fajr: raw.fajr, .sunrise: raw.sunrise, .dhuhr: raw.dhuhr,
            .asr: raw.asr, .maghrib: raw.maghrib, .isha: raw.isha,
        ]
        for (name, minutes) in settings.adjustments where minutes != 0 {
            times[name] = times[name]!.addingTimeInterval(TimeInterval(minutes * 60))
        }
        return PrayerDay(date: date, times: times)
    }

    /// Spike policy (M0): explicit rule from config wins; ≥48° with no explicit
    /// rule → seventh-of-the-night — stated directly (not via the library's
    /// `recommended`) so iOS and Android are identical by construction.
    static func effectiveHighLatitudeRule(settings: PrayerSettings, latitude: Double) -> String? {
        if let rule = settings.highLatitudeRule { return rule }
        guard abs(latitude) >= PrayerSettings.highLatitudeThreshold else { return nil }
        return "seventh_of_the_night"
    }

    /// Timeline for scheduling/widgets: `days` consecutive days from `startDate`.
    public func timeline(
        latitude: Double,
        longitude: Double,
        startDate: DateComponents,
        days: Int,
        settings: PrayerSettings,
        calendar: Calendar = .init(identifier: .gregorian)
    ) throws -> [PrayerDay] {
        guard let start = calendar.date(from: startDate) else { return [] }
        return try (0..<days).map { offset in
            let date = calendar.date(byAdding: .day, value: offset, to: start)!
            let components = calendar.dateComponents([.year, .month, .day], from: date)
            return try day(latitude: latitude, longitude: longitude, date: components, settings: settings)
        }
    }

    /// Next-prayer resolution across day boundaries. `today` and `tomorrow` must be
    /// consecutive (from `timeline`). Sunrise is never "next" — it is not a prayer.
    public static func nextPrayer(now: Date, today: PrayerDay, tomorrow: PrayerDay) -> NextPrayerState {
        let upcoming = today.ordered
            .filter { $0.name.isPrayer && $0.time > now }
        if let next = upcoming.first {
            let past = today.ordered.filter { $0.name.isPrayer && $0.time <= now }
            return NextPrayerState(current: past.last?.name, next: next.name, nextTime: next.time)
        }
        // After Isha: next is tomorrow's Fajr.
        return NextPrayerState(current: .isha, next: .fajr, nextTime: tomorrow.time(.fajr))
    }
}

/// Hijri date with the user/admin offset (ADR-0003: offset default is remote config).
public struct HijriDate: Equatable, Sendable {
    public let year: Int
    public let month: Int
    public let day: Int
    public let monthName: String

    public init(from date: Date, offsetDays: Int, locale: Locale = Locale(identifier: "ar")) {
        var calendar = Calendar(identifier: .islamicUmmAlQura)
        calendar.locale = locale
        let adjusted = calendar.date(byAdding: .day, value: offsetDays, to: date) ?? date
        let components = calendar.dateComponents([.year, .month, .day], from: adjusted)
        self.year = components.year ?? 0
        self.month = components.month ?? 0
        self.day = components.day ?? 0
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.locale = locale
        formatter.dateFormat = "MMMM"
        self.monthName = formatter.string(from: adjusted)
    }
}

File: android/core/prayer/src/main/kotlin/com/fatwabot/core/prayer/PrayerEngine.kt
package com.fatwabot.core.prayer

import java.time.LocalDate
import java.time.chrono.HijrahChronology
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlinx.datetime.Instant

/** User-facing prayer names. SUNRISE is displayed but is not a prayer. */
enum class PrayerNameUi(val isPrayer: Boolean) {
    FAJR(true), SUNRISE(false), DHUHR(true), ASR(true), MAGHRIB(true), ISHA(true)
}

/**
 * Resolved calculation settings — mirror of iOS PrayerKit.PrayerSettings.
 * Defaults flow: user override → /v1/config/prayer-defaults(country) → bundled fallback.
 */
data class PrayerSettings(
    // Umm al-Qura is the bundled default (stakeholder direction, 2026-07-12):
    // matches the Saudi authority users compare against, and is the offline
    // fallback whenever server config (which can override per country) is absent.
    val method: String = "umm_al_qura",
    val madhab: String = "shafi",
    /** null = automatic: recommended rule ≥48° (spike policy), none below. */
    val highLatitudeRule: String? = null,
    val adjustmentsMinutes: Map<PrayerNameUi, Int> = emptyMap(),
    val hijriOffsetDays: Int = 0,
) {
    val clampedAdjustments: Map<PrayerNameUi, Int>
        get() = adjustmentsMinutes.mapValues { it.value.coerceIn(-30, 30) }
    val clampedHijriOffset: Int
        get() = hijriOffsetDays.coerceIn(-2, 2)

    companion object {
        const val HIGH_LATITUDE_THRESHOLD = 48.0
    }
}

/** One civil day's times, adjustments applied. */
data class PrayerDayUi(
    val year: Int,
    val month: Int,
    val day: Int,
    val times: Map<PrayerNameUi, Instant>,
) {
    val ordered: List<Pair<PrayerNameUi, Instant>>
        get() = PrayerNameUi.entries.map { it to times.getValue(it) }
}

data class NextPrayerState(
    val current: PrayerNameUi?,
    val next: PrayerNameUi,
    val nextTime: Instant,
)

/** Deterministic engine over PrayerCalculator — mirror of iOS PrayerEngine. */
class PrayerEngine(private val calculator: PrayerCalculator = PrayerCalculator()) {

    fun day(
        latitude: Double,
        longitude: Double,
        year: Int,
        month: Int,
        dayOfMonth: Int,
        settings: PrayerSettings,
    ): PrayerDayUi {
        val rule = effectiveHighLatitudeRule(settings, latitude)
        val raw = calculator.times(
            PrayerCalculator.Request(
                latitude = latitude, longitude = longitude,
                year = year, month = month, day = dayOfMonth,
                method = settings.method, madhab = settings.madhab,
                highLatitudeRule = rule,
            ),
        )
        val base = mapOf(
            PrayerNameUi.FAJR to raw.fajr,
            PrayerNameUi.SUNRISE to raw.sunrise,
            PrayerNameUi.DHUHR to raw.dhuhr,
            PrayerNameUi.ASR to raw.asr,
            PrayerNameUi.MAGHRIB to raw.maghrib,
            PrayerNameUi.ISHA to raw.isha,
        )
        val adjustments = settings.clampedAdjustments
        val times = base.mapValues { (name, instant) ->
            val minutes = adjustments[name] ?: 0
            if (minutes == 0) instant else Instant.fromEpochSeconds(instant.epochSeconds + minutes * 60L)
        }
        return PrayerDayUi(year, month, dayOfMonth, times)
    }

    fun timeline(
        latitude: Double,
        longitude: Double,
        startYear: Int,
        startMonth: Int,
        startDay: Int,
        days: Int,
        settings: PrayerSettings,
    ): List<PrayerDayUi> {
        val start = LocalDate.of(startYear, startMonth, startDay)
        return (0 until days).map { offset ->
            val date = start.plusDays(offset.toLong())
            day(latitude, longitude, date.year, date.monthValue, date.dayOfMonth, settings)
        }
    }

    companion object {
        /**
         * Spike policy: explicit config rule wins; ≥48° with no rule → recommended,
         * never the library default. Recommendation logic mirrors adhan-swift's
         * HighLatitudeRule.recommended(for:): above 48° → seventh of the night.
         */
        fun effectiveHighLatitudeRule(settings: PrayerSettings, latitude: Double): String? {
            settings.highLatitudeRule?.let { return it }
            if (kotlin.math.abs(latitude) < PrayerSettings.HIGH_LATITUDE_THRESHOLD) return null
            return "seventh_of_the_night"
        }

        /** Next-prayer resolution across day boundaries; SUNRISE is never "next". */
        fun nextPrayer(now: Instant, today: PrayerDayUi, tomorrow: PrayerDayUi): NextPrayerState {
            val prayers = today.ordered.filter { it.first.isPrayer }
            val upcoming = prayers.firstOrNull { it.second > now }
            if (upcoming != null) {
                val current = prayers.lastOrNull { it.second <= now }?.first
                return NextPrayerState(current, upcoming.first, upcoming.second)
            }
            return NextPrayerState(
                PrayerNameUi.ISHA, PrayerNameUi.FAJR, tomorrow.times.getValue(PrayerNameUi.FAJR),
            )
        }
    }
}

/** Hijri date with offset — java.time HijrahChronology (Umm al-Qura). */
data class HijriDateUi(val year: Int, val month: Int, val day: Int, val monthName: String) {
    companion object {
        fun from(date: LocalDate, offsetDays: Int, locale: Locale = Locale("ar")): HijriDateUi {
            val hijrah = HijrahChronology.INSTANCE.date(date.plusDays(offsetDays.toLong()))
            val monthName = DateTimeFormatter.ofPattern("MMMM", locale).format(hijrah)
            return HijriDateUi(
                year = hijrah.get(java.time.temporal.ChronoField.YEAR),
                month = hijrah.get(java.time.temporal.ChronoField.MONTH_OF_YEAR),
                day = hijrah.get(java.time.temporal.ChronoField.DAY_OF_MONTH),
                monthName = monthName,
            )
        }
    }
}
